# Assignment 3

In this project, we will focus on healthcare. This data set is made available by MIT. It contains data about 9,026 heartbeat measurements. Each row represents a single measurement (captured on a timeline). There are a total of 80 data points (columns). This is a multiclass classification task: predict whether the measurement represents a normal heartbeat or other anomalies. 

# Read and Prepare the Data

In [2]:
# Importing Required packages

import numpy as np
import tensorflow as tf
from tensorflow import keras
import pandas as pd

In [3]:
# Reading data

heart_data = pd.read_csv("heartbeat_cleaned.csv")

In [4]:
# Seperating Target variable from Input variables

y = heart_data['Target']
x = heart_data.drop('Target', axis=1)

In [5]:
# Splitting data into test and train sets

from sklearn.model_selection import train_test_split

train_x, test_x, train_y, test_y = train_test_split(x, y, test_size=0.3)

### Data Transformation

In [6]:
#Target variables need to be an array with integer type

train_y = np.array(train_y)
test_y = np.array(test_y)

train_y = train_y.astype(int)
test_y = test_y.astype(int)


In [7]:
#Convert input variables to a 2-D array with float data type
train_x= np.array(train_x)
test_x= np.array(test_x)

train_x = train_x.astype(np.float32)
test_x = test_x.astype(np.float32)

In [8]:
#Keras expects a different input format:
#Data needs to have 3 dimensions

train_x = np.reshape(train_x, (train_x.shape[0], train_x.shape[1], 1))
test_x = np.reshape(test_x, (test_x.shape[0], test_x.shape[1], 1))


# Find the baseline

In [9]:
heart_data['Target'].value_counts()/len(heart_data)

0    0.582035
4    0.198995
2    0.155402
1    0.055905
3    0.007663
Name: Target, dtype: float64

#### The baseline shows the accuracy for each of our classifications from 0 to 4. As you can see the hieghest accuracy is for the first class with 58% accuracy. This means that our models should perform better than that.

# Build a cross-sectional shallow model using Keras (with only one hidden layer)

In [10]:
# importing early stopping from tenserflow
from tensorflow.keras.callbacks import EarlyStopping

earlystop = EarlyStopping(monitor='val_loss', patience=5, verbose=1, mode='auto')

callback = [earlystop]

In [11]:
model = keras.models.Sequential([
    
    keras.layers.Flatten(input_shape=[80, 1]),
    keras.layers.Dense(80, activation='relu'),
    keras.layers.Dense(5, activation='softmax')
    
])

In [12]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = tf.keras.optimizers.Nadam(learning_rate=0.01)

# If multiclass, use "sparse_categorical_crossentropy" as the loss function
model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])


history = model.fit(train_x, train_y, epochs=50,
                    validation_data=(test_x, test_y))

Epoch 1/50
175/175 [==============================] - 1s 2ms/step - loss: 0.6933 - accuracy: 0.7622 - val_loss: 1.0298 - val_accuracy: 0.6286
Epoch 2/50
175/175 [==============================] - 0s 1ms/step - loss: 0.4655 - accuracy: 0.8492 - val_loss: 0.4704 - val_accuracy: 0.8442
Epoch 3/50
175/175 [==============================] - 0s 1ms/step - loss: 0.3881 - accuracy: 0.8859 - val_loss: 0.3980 - val_accuracy: 0.8681
Epoch 4/50
175/175 [==============================] - 0s 1ms/step - loss: 0.3428 - accuracy: 0.9004 - val_loss: 0.3209 - val_accuracy: 0.9079
Epoch 5/50
175/175 [==============================] - 0s 1ms/step - loss: 0.3279 - accuracy: 0.9051 - val_loss: 0.2996 - val_accuracy: 0.9100
Epoch 6/50
175/175 [==============================] - 0s 1ms/step - loss: 0.3064 - accuracy: 0.9113 - val_loss: 1.4346 - val_accuracy: 0.7186
Epoch 7/50
175/175 [==============================] - 0s 1ms/step - loss: 0.3063 - accuracy: 0.9121 - val_loss: 0.3875 - val_accuracy: 0.8760
Epoch 

In [13]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.5477070808410645, 0.876884400844574]

In [14]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.55
accuracy: 87.69%


### Our first model is shallow cross sectional with one hidden layer. this model accuracy is 93.72 percent using relu as hidden layers activation function and softmax as output layer activation function (Because our out put is multi class classification) which is almost 2 times better than the base line. we also used selu as activation function but the result was less accurate.

# Build a cross-sectional deep model using Keras (with two or more hidden layers):

In [15]:
model = keras.models.Sequential([
    
    keras.layers.Flatten(input_shape=[80, 1]),
    keras.layers.Dense(80, activation='relu'),
    keras.layers.Dense(80, activation='relu'),
    keras.layers.Dense(80, activation='relu'),
    keras.layers.Dense(5, activation='softmax')
    
])

In [16]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = tf.keras.optimizers.Nadam(learning_rate=0.01)

# If multiclass, use "sparse_categorical_crossentropy" as the loss function
model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])


history = model.fit(train_x, train_y, epochs=50,
                    validation_data=(test_x, test_y))

Epoch 1/50
175/175 [==============================] - 2s 4ms/step - loss: 0.7367 - accuracy: 0.7434 - val_loss: 1.0474 - val_accuracy: 0.6022
Epoch 2/50
175/175 [==============================] - 1s 3ms/step - loss: 0.4698 - accuracy: 0.8388 - val_loss: 0.4152 - val_accuracy: 0.8647
Epoch 3/50
175/175 [==============================] - 0s 3ms/step - loss: 0.3950 - accuracy: 0.8726 - val_loss: 0.3335 - val_accuracy: 0.9016
Epoch 4/50
175/175 [==============================] - 1s 3ms/step - loss: 0.3446 - accuracy: 0.8925 - val_loss: 0.3451 - val_accuracy: 0.8894
Epoch 5/50
175/175 [==============================] - 1s 3ms/step - loss: 0.3153 - accuracy: 0.9042 - val_loss: 0.3223 - val_accuracy: 0.8924
Epoch 6/50
175/175 [==============================] - 1s 3ms/step - loss: 0.3011 - accuracy: 0.9070 - val_loss: 0.5408 - val_accuracy: 0.8254
Epoch 7/50
175/175 [==============================] - 1s 3ms/step - loss: 0.2821 - accuracy: 0.9139 - val_loss: 0.3132 - val_accuracy: 0.8978
Epoch 

In [17]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.3435995578765869, 0.9237855672836304]

In [18]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.34
accuracy: 92.38%


### In this model we used keras and build a model with 3 hidden layers and activation function of relu. The output layer is still softmax due to multi class classification. we increased the number of the hidden layers and get better results infact we got 94 percent accuracy which is a little bit better than model with one layer.

# Build a sequential shallow LSTM Model (with only one LSTM layer):

In [19]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
        keras.layers.LSTM(80, activation='softmax' , input_shape=[n_steps, n_inputs])
])

In [20]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=50,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/50
175/175 [==============================] - 15s 72ms/step - loss: 2.1384 - accuracy: 0.5698 - val_loss: 1.1784 - val_accuracy: 0.5490
Epoch 2/50
175/175 [==============================] - 14s 80ms/step - loss: 1.0816 - accuracy: 0.6023 - val_loss: 1.0799 - val_accuracy: 0.5586
Epoch 3/50
175/175 [==============================] - 14s 79ms/step - loss: 1.0576 - accuracy: 0.6019 - val_loss: 1.0614 - val_accuracy: 0.6219
Epoch 4/50
175/175 [==============================] - 13s 77ms/step - loss: 1.0354 - accuracy: 0.6136 - val_loss: 1.0585 - val_accuracy: 0.6214
Epoch 5/50
175/175 [==============================] - 14s 77ms/step - loss: 1.0185 - accuracy: 0.6280 - val_loss: 1.0175 - val_accuracy: 0.6244
Epoch 6/50
175/175 [==============================] - 13s 77ms/step - loss: 0.9906 - accuracy: 0.6441 - val_loss: 1.0070 - val_accuracy: 0.6080
Epoch 7/50
175/175 [==============================] - 13s 76ms/step - loss: 0.9713 - accuracy: 0.6455 - val_loss: 1.0176 - val_accuracy:

In [21]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.7649096250534058, 0.7349246144294739]

In [22]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.76
accuracy: 73.49%


### In our LSTM model with only one layer we achieved the accuracy of 79 percent which is somehow better than base line but way worse than our previous models.

# Build a sequential deep LSTM Model (with only two LSTM layers):

In [23]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.LSTM(80, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.LSTM(80),
    keras.layers.Dense(5, activation='softmax')
])

In [24]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=50,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/50
175/175 [==============================] - 25s 131ms/step - loss: 1.1360 - accuracy: 0.5824 - val_loss: 1.5904 - val_accuracy: 0.2044
Epoch 2/50
175/175 [==============================] - 22s 123ms/step - loss: 1.1016 - accuracy: 0.5896 - val_loss: 1.1585 - val_accuracy: 0.5691
Epoch 3/50
175/175 [==============================] - 23s 129ms/step - loss: 1.1294 - accuracy: 0.5876 - val_loss: 1.1567 - val_accuracy: 0.5691
Epoch 4/50
175/175 [==============================] - 23s 132ms/step - loss: 1.1272 - accuracy: 0.5876 - val_loss: 1.1861 - val_accuracy: 0.5691
Epoch 5/50
175/175 [==============================] - 22s 127ms/step - loss: 1.1278 - accuracy: 0.5876 - val_loss: 1.1519 - val_accuracy: 0.5691
Epoch 6/50
175/175 [==============================] - 21s 119ms/step - loss: 1.1264 - accuracy: 0.5876 - val_loss: 1.1921 - val_accuracy: 0.5691
Epoch 7/50
175/175 [==============================] - 21s 120ms/step - loss: 1.1263 - accuracy: 0.5876 - val_loss: 1.2104 - val_ac

In [25]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[1.1576522588729858, 0.5690954923629761]

In [26]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 1.16
accuracy: 56.91%


### By trying to add more hidden layers to LSTM model the accuracy dropped even further down to 58 percent.

# Build a sequential shallow GRU Model (with only one GRU layer):

In [27]:
n_steps = 80
n_inputs = 1

model = keras.models.Sequential([
    keras.layers.GRU(80, input_shape=[n_steps, n_inputs]),
    keras.layers.Dense(5, activation='softmax')
])

In [28]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=50,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/50
175/175 [==============================] - 14s 58ms/step - loss: 1.0688 - accuracy: 0.6030 - val_loss: 1.1869 - val_accuracy: 0.5117
Epoch 2/50
175/175 [==============================] - 9s 54ms/step - loss: 0.8402 - accuracy: 0.6965 - val_loss: 0.6008 - val_accuracy: 0.7952
Epoch 3/50
175/175 [==============================] - 10s 56ms/step - loss: 0.6821 - accuracy: 0.7737 - val_loss: 0.7512 - val_accuracy: 0.7508
Epoch 4/50
175/175 [==============================] - 9s 54ms/step - loss: 0.5402 - accuracy: 0.8324 - val_loss: 0.4887 - val_accuracy: 0.8492
Epoch 5/50
175/175 [==============================] - 9s 53ms/step - loss: 0.4790 - accuracy: 0.8489 - val_loss: 0.4182 - val_accuracy: 0.8614
Epoch 6/50
175/175 [==============================] - 9s 52ms/step - loss: 0.4416 - accuracy: 0.8613 - val_loss: 0.5613 - val_accuracy: 0.8392
Epoch 7/50
175/175 [==============================] - 10s 54ms/step - loss: 0.4061 - accuracy: 0.8728 - val_loss: 0.4496 - val_accuracy: 0.8

In [29]:
# evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

[0.25891128182411194, 0.9216917753219604]

In [30]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


loss: 0.26
accuracy: 92.17%


### In our GRU model with only one layer the accuracy is 91 percent which is better than baseline and LSTM with one layer but still not as good as first two models. By increasing the nuber of epochs model performed wose

# Build a sequential deep GRU Model (with only two GRU layers):

In [31]:
model = keras.models.Sequential([
    keras.layers.GRU(80, return_sequences=True, input_shape=[n_steps, n_inputs]),
    keras.layers.GRU(80),
    keras.layers.Dense(5, activation='softmax')
])

In [ ]:
np.random.seed(42)
tf.random.set_seed(42)

optimizer = keras.optimizers.Nadam(learning_rate=0.01)

model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=['accuracy'])

history = model.fit(train_x, train_y, epochs=50,
                   validation_data = (test_x, test_y), callbacks=callback)

Epoch 1/50
175/175 [==============================] - 28s 118ms/step - loss: 0.9729 - accuracy: 0.6405 - val_loss: 1.2436 - val_accuracy: 0.4920
Epoch 2/50
175/175 [==============================] - 18s 104ms/step - loss: 0.4995 - accuracy: 0.8413 - val_loss: 0.7619 - val_accuracy: 0.7688
Epoch 3/50
175/175 [==============================] - 18s 105ms/step - loss: 0.3866 - accuracy: 0.8810 - val_loss: 0.3334 - val_accuracy: 0.8961
Epoch 4/50
175/175 [==============================] - 19s 107ms/step - loss: 0.3773 - accuracy: 0.8841 - val_loss: 0.2998 - val_accuracy: 0.9020
Epoch 5/50
175/175 [==============================] - 19s 108ms/step - loss: 0.3062 - accuracy: 0.9056 - val_loss: 0.3126 - val_accuracy: 0.9016
Epoch 6/50
175/175 [==============================] - 19s 106ms/step - loss: 0.2691 - accuracy: 0.9162 - val_loss: 1.1549 - val_accuracy: 0.7425
Epoch 7/50
175/175 [==============================] - 19s 109ms/step - loss: 0.2980 - accuracy: 0.9058 - val_loss: 0.3253 - val_ac

In [ ]:
 #evaluate the model

scores = model.evaluate(test_x, test_y, verbose=0)

scores

# In results, first is loss, second is accuracy

In [ ]:
# extract the accuracy from model.evaluate

print("%s: %.2f" % (model.metrics_names[0], scores[0]))
print("%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))


### By adding one more hidden layer to the model the GRU it doesn't performs better even though it takes longer to process. the accuracy of 84 % . By increasing the nuber of epochs model performed abit better

## List the test values of each model